src/node_watchdog.h

```diff
@@ -54,24 +61,54 @@ class Watchdog {
   bool* timed_out_;
 };
 
-class SigintWatchdog {
+class SigintWatchdogBase {
+ public:
+  virtual ~SigintWatchdogBase() = default;
+  virtual SignalPropagation HandleSigint() = 0;
+};
+
+class SigintWatchdog : public SigintWatchdogBase {
  public:
   explicit SigintWatchdog(v8::Isolate* isolate,
                           bool* received_signal = nullptr);
   ~SigintWatchdog();
   v8::Isolate* isolate() { return isolate_; }
-  void HandleSigint();
+  SignalPropagation HandleSigint() override;
 
  private:
   v8::Isolate* isolate_;
   bool* received_signal_;
 };
 
+class TraceSigintWatchdog : public HandleWrap, public SigintWatchdogBase {
+ public:
+  static void Init(Environment* env, Local<v8::Object> target);
+  static void New(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Start(const v8::FunctionCallbackInfo<Value>& args);
+  static void Stop(const v8::FunctionCallbackInfo<Value>& args);
+
+  SignalPropagation HandleSigint() override;
+
+  SET_NO_MEMORY_INFO()
+  SET_MEMORY_INFO_NAME(TraceSigintWatchdog)
+  SET_SELF_SIZE(TraceSigintWatchdog)
+
+ private:
+  enum SignalFlags { None, FromIdle, FromInterrupt };
```

Comment:
Since you already use the namespace everywhere..

```suggestion
  enum class SignalFlags { None, FromIdle, FromInterrupt };
```